logic/scorer/scorer.go

```diff
@@ -60,260 +34,247 @@ func points(lead model.Card, hand []model.Card, isCrib bool) int {
 	}
 
-	var st scoreType
-	pts := 0
+	if numFifteens > 0 {
+		return fifteen0 << numFifteens, numFifteens * 2
+	}
+	return 0, 0
+}
 
-	if ((min+1 == values[1]) && (values[1]+1 == values[2])) ||
-		((values[2]+1 == values[3]) && (values[3]+1 == max)) {
-		st = st | run3 /* run of 3 (and maybe a pair) */
-		pts += 3
+func howManyAddUpTo(goal int, ptVals []int) int {
+	if len(ptVals) == 0 {
+		return 0
 	}
 
-	if numPairs == 1 {
-		// we already checked for two above
-		st = st | onepair /* one pair */
-		pts += 2
+	var many int
+	for i, o := range ptVals {
+		if o == goal {
+			many++
+		} else if o < goal {
+			// o is less than the goal. See what we can find with it
+			subWith := howManyAddUpTo(goal-o, ptVals[i+1:])
+			many += subWith
+		}
 	}
 
-	return st, pts
+	return many
 }
 
-func scoreFlushesAndNobs(lead model.Card, hand []model.Card, isCrib bool) (scoreType, int) {
-	var st scoreType
-	pts := 0
-	handSuit := hand[0].Suit
-	isHandFlush := true
-	for _, c := range hand {
-		if c.Suit != handSuit {
-			isHandFlush = false
+func scoreRuns(values [numCardsToScore]int, valuesToCounts []uint8) (scoreType, int) {
+	for _, v := range values {
+		if valuesToCounts[v-1] > 0 {
+			// this is already part of a run; skip calculation
+			continue
 		}
-		if c.Value == 11 /* Jack */ && c.Suit == lead.Suit {
-			st = st | nobs
-			pts++
+		runLen := uint8(1)
+		mult := valuesToCounts[v]
+		// we're at the potential beginning of a run
+		for next := v + 1; valuesToCounts[next] > 0; next++ {
+			mult *= valuesToCounts[next]
+			runLen++
+		}
+		if runLen >= 3 {
+			return calculateTypeAndPoints(runLen, mult)
 		}
 	}
+	return 0, 0
+}
 
-	if isHandFlush {
-		if lead.Suit == handSuit {
-			st = st | flush5
-			pts += 5
-		} else if !isCrib {
-			st = st | flush4
-			pts += 4
-		}
+func calculateTypeAndPoints(longest, mult uint8) (scoreType, int) {
+	if longest == 5 {
+		return run5, 5
 	}
+	// typeMap maps the run length and multiplier to a scoring type
+	typeMap := [5][5]scoreType{
+		3: {1: run3, 2: doubleRunOfThree, 3: tripleRunOfThree, 4: doubleDoubleRunOfThree},
+		4: {1: run4, 2: doubleRunOfFour},
+	}
+	return typeMap[longest][mult], int(longest) * int(mult)
+}
+
+func resolveScoreType(st scoreType) scoreType {
+	// these masks are generated as following:
+	// 00011111
+	//    ^ doubleRunOfThree bit
+	// 00000011
+	//      ^ tripleRunOfThree bit (minus one to get the two bits below)
+	// they include the bits set for th
+	runsWithPairsMask := generateBitMask(tripleRunOfThree, doubleRunOfThree)
+	if runsWithPairs := runsWithPairsMask & st; runsWithPairs > 0 {
+		// we have a run that has a pair baked in, like tripleRunOfThree
+		// let's mask off quad, triplet, onepair, and twopair
+		pairsMask := generateBitMask(quad, twopair)
+		return st &^ pairsMask
+	}
+	return st
+}
 
-	return st, pts
+// generateBitMask generates a bit mask with the bits from min to max (inclusive) set and the rest unset
+func generateBitMask(min, max scoreType) scoreType {
```

Comment:
:mag: is this necessary? Could we get away with defining another set of consts `runsWithSomePairs` and `allPairingScores`?